### ISSUE TITLE
When cursor moves horizontally, it shows two different behaviors when passing `inlayHints`

### ISSUE THREAD

**Issue by brandonwie** [microsoft/vscode#166927]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.73.1
- OS Version: Ventura 13.0.1

> the screenshots were taken with `code --disabled-extensions`

### A. Behavior when the cursor moves forward (left to right),
<img width="381" alt="when cursor moves forward with inlayHints options 01" src="https://user-images.githubusercontent.com/31783269/203210540-f88ca65b-a9eb-48e7-9266-65ba3bf39f0e.png">
<img width="386" alt="when cursor moves forward with inlayHints options 02" src="https://user-images.githubusercontent.com/31783269/203210567-b2430db7-2ca9-4153-896c-fdc738899d42.png">
<img width="380" alt="when cursor moves forward with inlayHints options 03" src="https://user-images.githubusercontent.com/31783269/203210591-2e010d7d-14d2-4770-90ed-5675cb73701f.png">
As you can see above, the cursor does not pass inlayHints instead of going to the right in front of the letter A of the second `Array`.
However, when you move the cursor in the opposite direction, from right to left, it shows different behavior.
---
### B. Behavior when the cursor moves backward (right to left),
<img width="382" alt="when cursor moves backward with inlayHints options 01" src="https://user-images.githubusercontent.com/31783269/203211565-2131c34c-80a8-4a69-8928-a4e83c8698c5.png">
<img width="381" alt="when cursor moves backward with inlayHints options 02" src="https://user-images.githubusercontent.com/31783269/203211631-cb646f60-5814-4f31-912b-4774bf045a67.png">
<img width="388" alt="when cursor moves backward with inlayHints options 03" src="https://user-images.githubusercontent.com/31783269/203211667-05a009a8-e760-4ec6-8515-63b694685c77.png">
The cursor stays in front of the letter A on the first move, but on the next move, it passes the opening bracket and it locates before the opening bracket. (This pattern should be applied to the A case for consistency.)

### Conclusion
When a cursor moves horizontally, the cursor ignores text that is placed right after the inlay hints.
> from the screenshots
> left to right: letter `A` is ignored because it is placed right after the inlay hint
> right to left: the opening bracket is ignored because it is also placed right after the inlay hint

### Suggestions
- the inlay hints are literally the 'hints' not 'texts' so the cursor should not be blocked by it as you can observe in case A (the behavior in case B should apply to the A case as well)

Steps to Reproduce:

1. turn on `inlayHints` options by putting these lines in settings.json of your VSCode
```json
{
  // js ts type annotations
  "javascript.inlayHints.parameterNames.enabled": "all",
  "typescript.inlayHints.parameterNames.enabled": "all",
  "javascript.inlayHints.variableTypes.enabled": true,
  "typescript.inlayHints.variableTypes.enabled": true,
  "javascript.inlayHints.propertyDeclarationTypes.enabled": true,
  "typescript.inlayHints.propertyDeclarationTypes.enabled": true,
  "javascript.inlayHints.parameterTypes.enabled": true,
  "typescript.inlayHints.parameterTypes.enabled": true,
  "javascript.inlayHints.functionLikeReturnTypes.enabled": true,
  "typescript.inlayHints.functionLikeReturnTypes.enabled": true,
}
```
2. create `test.(js | ts)` file 
3. write any function that has parameters and return value
4. you see inlayHints displayed
5. put your cursor before or after the inlayHints and move your cursor
6. observe how it behaves differently while moving the cursor left and right horizontally

**Comment by jrieken** [comment]
@hediet I have to agree with @brandonwie that this is confusing and we should at least consider an option for this

**Comment by brandonwie** [comment]
@hediet @jrieken thank you so much you guys for the comments
I think the "design" can be categorized in two different ways, if I may, either it is conceptual or functional.

In a conceptual way, the cursor behaves the **same way** with the `inlayHints` as "When a cursor moves horizontally, the cursor ignores text that is placed right after the inlay hints."
It totally makes sense that a function(in this case, cursor and inlay hint) returns the same output(behavior) when the inputs(a cursor passing an inlay hint) are the same. (but maybe if the function's returning concept has one more condition, "the inlay hint should not be treated as text", then maybe we can see it from a different angle)

However, in a functional manner, it is very confusing as jrieken mentioned. One of the reasons is that when the cursor passes the inlay hint from left to right, it gives me the impression that the inlay hint is treated as text because the cursor stops in front of it.  I truly love and use this brilliant feature every single day, and the behavior was one of the things I need to endure to use the feature. However, after months, I have come to a point where I want to raise my hand and ask you guys if there is any chance that the behavior can be changed in a more practical manner.

I'm more than happy to know the fact that it was by design. Either way, I'll keep using it. 😁
Thanks a lot.

**Comment by ifyuhj** [comment]
@brandonwie  I fully support your viewpoint, it is well-reasoned and convincing.
> Suggestions
the inlay hints are literally the 'hints' not 'texts' so the cursor should not be blocked by it as you can observe in case A (the behavior in case B should apply to the A case)

Your suggestions will fix this: 
> Set the cursor shape to block(the same as the normal mode in Vim), when the cursor moves forward (left to right), it can stop in inlay hint.
<img width="409" alt="image" src="https://user-images.githubusercontent.com/59051928/212533815-0828a54a-870a-44d9-83c6-871800d2eca3.png">

Although the current design should be well thought out, it wouldn't hurt to give users an additional option.
In any case, I am grateful to the developers of VSCode for their contributions, giving developers a free and powerful development tool.

**Comment by andreasvh-conceto** [comment]
Hi @jrieken @hediet 

Could this be related to this issue? 
https://github.com/microsoft/vscode/pull/226515#issuecomment-2466212707

The PR from @NriotHrreion is already there to be reviewed since several months related to the confusing and unacceptable behaviour of the block cursor with active inlay hints. Maybe it will also fix this issue here, which is also 2 years old.

@NriotHrreion provided a fix for that. Would appreciate to be merged soon or at least a statement when we can expect this to be fixed.

Thanks a lot and best regards.
Andreas
